plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -391,4 +386,97 @@ public int read(byte[] b, int off, int len) throws IOException {
                 "PUT_BLOCK", putBlockOperations.get());
         }
     }
+
+    /**
+     * Building our own input stream instead of using the SDK's {@link com.microsoft.azure.storage.blob.BlobInputStream}
+     * because that stream is highly inefficient in both memory and CPU use.
+     */
+    private static class BlobInputStream extends InputStream {
```

Comment:
Technically speaking, we could avoid this whole dance and make the Azure download as fast as those in GCS and S3 by offering a different blob container API that consumes an `OutputStream`, that would do away with the chunked reading and make things fast (far as I can tell, haven't tried it yet). An annoying hack for sure, but probably less effort+risk than upgrading to SDK v12. For now this is the best I could come up with from the profiling @hub-cap did.

Reply:
I did some work to upgrade to SDK v12, but since it wasn't a priority I've been working on other things. I did some small benchmarks and the difference was notable, we can revisit the priority for future releases as all the past blockers are solved with the latest SDK versions.

Reply:
Yea this certainly isn't meant as a long term solution. Just a stop-gap to :)